chore: blocking user spec [19940]

JacquelineLehner opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-19940" title="WPB-19940" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-19940</a>  [Web/QA] Write the block regression tests in Playwright
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->

## Description
Add regression tests for blocking users in eight cases

<!-- Uncomment this section if your PR has UI changes -->
<!--
## Screenshots/Screencast (for UI changes)
-->

## Checklist

- [x] mentions the JIRA issue in the PR name ([WPB-19940]
- [x] PR has been self reviewed by the author;
- [x] Hard-to-understand areas of the code have been commented;
- [ ] If it is a core feature, unit tests have been added;

<!-- Uncomment this section if it is necessary to understand the PR -->
<!-- ## Important Details for the Reviewers

- use (x) data
- can be reviewed commit-by-commit
- be sure to look at ... -->


[WPB-19940]: https://wearezeta.atlassian.net/browse/WPB-19940?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

xdubx committed:
test(WPB-19940): add tests for blocking users

markbrockhoff commented on test/e2e_tests/specs/Folders/Folders.spec.ts:
```diff
@@ -0,0 +1,208 @@
+let userA: User;
+let userB: User;
+
+test.describe('Folders', () => {
```
Should the folder tests really be part of this PR? Its name suggests something else, maybe they should be split into their own PR? (I'm skipping the review of this file for now as I feel like it should be part of a different PR)

markbrockhoff commented on test/e2e_tests/pageManager/webapp/modals/conversationNotConnected.modal.ts:
```diff
@@ -0,0 +1,28 @@
+
+import {BaseModal} from './base.modal';
+
+export class ConversationNotConnectedModal extends BaseModal {
```
This POM is identical to the already existing `AcknowledgeModal` (see: `test/e2e_tests/pageManager/webapp/modals/acknowledge.modal.ts`). Since I can't seem to find a dedicated ConversationNotConnected Modal component this POM would match we should not create a duplicate abstraction for it :)

markbrockhoff commented on test/e2e_tests/pageManager/webapp/modals/createNewFolder.modal.ts:
```diff
@@ -0,0 +1,35 @@
+  readonly folderNameInput: Locator;
+
+  constructor(page: Page) {
+    super(page, 'modal-template-option');
```
Are you sure the modal shown here is of type "modal-template-option"? Looking into the function where it's opened it rather looks like a modal of type "input" to me. (Source: src/script/repositories/conversation/ConversationLabelRepository.ts - addConversationToNewLabel)

To explain:
Almost all modals used throughout the application are just sub-versions of the so called `PrimaryModal`. This modal is opened in different variations like `OPTION` which will show the cancel and do action buttons.
Ideally we don't want the POMs (Page Object Models) like this one to macht the real pages / components in the application. In the application there's no "CreateNewFolderModal" but there are multiple occurrences of an option modal just with different titles. So instead of creating a new modal POM for every opened modal, let's try to stick to the kinds of modals which can exist.
You can see all types of modals by looking at the enum `PrimaryModalType` in PrimaryModalTypes.ts

markbrockhoff commented on test/e2e_tests/pageManager/webapp/modals/unblockWarning.modal.ts:
```diff
@@ -0,0 +1,64 @@
+    this.cancelButton = this.modal.locator("[data-uie-name='do-secondary']");
+  }
+
+  async isModalPresent() {
```
Please avoid such "is*" utils on POMs. They will require you to write the expect for it like `expect(await is*()).toBe(true);` which is a bad practice in Playwright. Instead we want to use native assertions like `await expect(something).toBeVisible()`.
To archive that you can just use the locator already available in here instead of creating a function wrapping it, e.g.:
`await expect(modals.unblockWarningModal.modal).toBeVisible()`

markbrockhoff commented on test/e2e_tests/pageManager/webapp/pages/conversationList.page.ts:
```diff
@@ -50,6 +59,15 @@ export class ConversationListPage {
     this.unarchiveConversationMenuButton = page.locator(selectById('btn-unarchive'));
+    this.moveConversationToFolderButton = page.locator(selectById('btn-move-to'));
+    this.createNewFolderButton = page.locator(selectById('btn-create-new-folder'));
+    this.customFolderButton = page.locator(selectByDataAttribute('conversation-label-context-menu'));
```
Are there no better selectors for the buttons than their id? Usually depending on the id of an element is not a really good practice within tests

markbrockhoff commented on test/e2e_tests/specs/Block/block2.spec.ts:
```diff
@@ -0,0 +1,55 @@
+
+test.afterAll(async ({api}) => {
+  await tearDownAll(api);
+});
```
One of the rare occasions a copilot comment is actually right xD
We only added block2.spec.ts while trying out things. Please remove it from the PR, if anything in here is still relevant it should be moved over to block.spec.ts :)

markbrockhoff commented on test/e2e_tests/utils/userActions.ts:
```diff
@@ -112,3 +113,95 @@ export const handleAppLockState = async (pageManager: PageManager, appLockPassCo
+/**
+ * Logs in a user and handles initial setup
+ */
+export async function loginAndSetup(user: User, pageManager: PageManager) {
```
Please prefer the fixtures `createUser()` and `createPage()` with the plugins `withLogin` instead of creating a new util. You can find examples on how to use them in `edit.spec.ts`

markbrockhoff commented on test/e2e_tests/utils/userActions.ts:
```diff
@@ -112,3 +113,95 @@ export const handleAppLockState = async (pageManager: PageManager, appLockPassCo
+  await userAPages.startUI().selectUser(userB.username);
+  await userAModals.userProfile().clickConnectButton();
+
+  expect(await userAPages.conversationList().isConversationItemVisible(userB.fullName)).toBeTruthy();
```
Copilot is right about the await but that's just because you tried to re-use the `isConversationItemVisible` function which is one of the "is*" functions we'd like to slowly get rid of.
Generally speaking I'd advise against using an `expect` within a util. The util should only perform the stated action, waiting before continuing is not part of it's responsibility. Either the test should do the expect after triggering the action, or the action performed after it should be stable enough to only continue once ready.

markbrockhoff commented on test/e2e_tests/utils/userActions.ts:
```diff
@@ -112,3 +113,95 @@ export const handleAppLockState = async (pageManager: PageManager, appLockPassCo
+  await modals.blockWarning().clickBlock();
+
+  // Optional handling for modals that appear after blocking
+  if (handleUnableToOpenModal) {
```
This sound's more like the workaround for a bug than a feature to me xD
Is there really a reasonable situation in which a modal should pop up telling you you can't open a conversation which you're already on and just blocked?

markbrockhoff commented on test/e2e_tests/utils/userActions.ts:
```diff
@@ -112,3 +113,95 @@ export const handleAppLockState = async (pageManager: PageManager, appLockPassCo
+ * @param pageManager PageManager of the blocking user
+ * @param userToBlock User object of the user to be blocked
+ */
+export async function blockUserFromProfileView(pageManager: PageManager, userToBlock: User) {
```
Are you sure it's reasonable to already abstract this function even though it is only used once? Not sure but it doesn't sound to me like it would be re-used a lot, so maybe it would be better to just keep it as a step in the test for now. What do you think?

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+    );
+
+    test.afterEach(async ({api}) => {
+      await tearDownAll(api);
```
FYI, once #19804 is merged, there will be a fixture available for setting up teams ;)

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+        // Step 4: User A clicks 'Cancel' button
+        await userAModals.blockWarning().clickCancel();
+        // Step 5: Conversation is still present, and User A can open it
+        await userAPages.conversationList().openConversation(userB.fullName);
```
This test doesn't have any assertion, you just rely on the call to openConversation to not fail but this is not a valid way of testing the behavior. For example: If you change line 104 to call `clickBlock()` instead of `clickCancel()` the test still passes.

Tip: Always try to make your test fail once before making it pass ;)

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+        // Step 6: User B cannot send message to User A
+        await userBPages.conversationList().openConversation(userA.fullName);
+        await userBPages.conversation().sendMessage('Message after block');
+        expect(await userAPages.conversation().messageCount()).toBe(0);
```
Hint: A more reliable way of checking the number of messages would be:
```suggestion
        await expect(userAPages.conversation().messages).toHaveCount(0);
```

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+          await userBPages.conversationList().openConversation(conversationName);
+        });
+
+        // Step 1: User B sends message to group chat with User A
```
[Nitpick] Maybe it's sufficient to only have a comment or a step with a description and not both ;)

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+
+          // Step 1: User A sends connection request to User B
+          await test.step('Preconditions: User A connects with User B', async () => {
+            await connectUsersManually(userA, userB, userAPageManager, userBPageManager, false);
```
Are you sure this util is the right one to use here? To me the test title reads like the user should be blocked before the connection request is accepted, otherwise there wouldn't be a difference to previous tests or would it?

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+          // Step 2: User A blocks User B from conversation list
+          await blockUserFromConversationList(userAPageManager, userB);
+          // Step 3: User B does not see Connection Request from User A
+          expect(userBPages.conversationList()).not.toContain(userA.fullName);
```
Copilot is right, just wanted to point out the same thing. toContain will check if the given substring is part of the string within `expect`. But conversationList returns a class instance so this will always be true. E.g. you can comment out line 255 and not block the user and the test would still pass.

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+      'Verify you can unblock someone from conversation list options',
+      {tag: ['@TC-148', '@regression']},
+      async ({pageManager: userAPageManager, userBPageManager}) => {
+        test.slow();
```
I don't think is is necessary now that we increased the default test timeout to 90s. On my maschine the test takes 29s so no need to increase the timeout to 270s ;)

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+          await userBPages.conversationList().openConversation(userA.fullName);
+          await userBPages.conversation().sendMessage('Message after being unblocked');
+          await userAPages.conversationList().openConversation(userB.fullName);
+          expect(await userBPages.conversation().messageCount()).toBe(2);
```
Copilot is right

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+      userB = getUser();
+      const teamName = 'Team Name';
+      const context = await browser.newContext();
+      const page = await context.newPage();
```
Is it really necessary to open a new page just to log in userB and then do nothing with it? I was able to remove the whole context without the test failing.

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+        // Step 2: User A opens the options menu for user B
+        await userAPages.conversationList().clickConversationOptions(userB.fullName);
+
+        await expect(userAPages.conversationList().blockConversationMenuButton).not.toBeAttached();
```
That's a nice assertion 💪🏻

JacquelineLehner committed:
test(block): refactor block.spec.ts according to Review-Comments

add function to userActions.ts to send a connection request

add fixture for connection request

delete conversationNotConnected.modal.ts, replace it with acknowledge.modal.ts

JacquelineLehner commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -0,0 +1,344 @@
+    });
+
+    test(
+      'I want to cancel blocking a 1on1 conversation from conversation list 0',
```
fixed typo

markbrockhoff commented on test/e2e_tests/pageManager/webapp/pages/conversationList.page.ts:
```diff
@@ -123,4 +135,15 @@ export class ConversationListPage {
   }
+
+  async getUserAvatarWrapper(user: User): Promise<Locator> {
+    return this.page
```
Question: Could you maybe reuse the `getConversationLocator()` function in here to get the conversation itself and then only locate the avatar within it?

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -104,222 +103,241 @@ test.describe('Block', () => {
+        await userAPages.conversationList().openConversation(userB.fullName);
+        await expect(userAPages.conversation().messageInput).not.toBeVisible();
+        // Step 6: User B cannot send message to User A
+        await userBPages.conversationList().openConversation(userA.fullName);
```
Checking for the sent message to not be received isn't really reliable, since it creates a race condition where the test could pass simply because the message way not yet received.
My general question to you would be why you even send a message here. I appreciate the effort but checking that the message from a blocked user is not received is not part of the test's scope. My suggestion would by to ensure the user is blocked like in the test above :)

JacquelineLehner replied:
Assertion is changed now. The messages between A and B are required from the Testiny Specifications.

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -104,222 +103,241 @@ test.describe('Block', () => {
+      'Verify you still receive messages from blocked person in a group chat 0',
+      {tag: ['@TC-141', '@regression']},
+      async ({createPage}) => {
+        const userAContext = await PageManager.from(createPage(withLogin(userA), withConnectionRequest(userB)));
```
Why is this constant named `userAContext`? It's not a browser context but a PageManager instance

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -104,222 +103,241 @@ test.describe('Block', () => {
+
+        // Step 1: User B blocks User A
+        await test.step('User B blocks User A', async () => {
+          await userAPages.conversationList().openConversation(userC.fullName);
```
Could you please add a comment explaining why userA opens a conversation with userC here? I know that it's to avoid the can't open conversation modal since we looked at that together, but others don't :)

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -104,222 +103,241 @@ test.describe('Block', () => {
+
+        // Step 2: User A wants to add B to a group chat after being blocked by User B
+        await test.step('Users A tries to add B to a group', async () => {
+          await userAComponents.conversationSidebar().isPageLoaded();
```
I'm pretty sure the page is loaded at this point of the test or is there a valid reason to manually check for it here?

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -104,222 +103,241 @@ test.describe('Block', () => {
+          await userAPages.startUI().selectUsers([userB.username]);
+          await userAPages.groupCreation().clickCreateGroupButton();
+
+          // Step 3: Modal 'modalConversationNotConnectedMessageOne' is visible
```
[Nitpick] I think this comment is rather misleading due to its "Step 3:" prefix within Step 2

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -104,222 +103,241 @@ test.describe('Block', () => {
+        // Step 2: User A blocks User B from conversation list
+        await blockUserFromConversationList(userAContext, userB);
+        // Step 3: User B does not see Connection Request from User A
+        await expect(userBPages.conversationList().getConversationLocator(userA.fullName)).not.toBeVisible();
```
This test passes even if the action above is commented out. Please make sure the test can fail, e.g. by making sure the conversation is visible for userB before A blocks him.

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -104,222 +103,241 @@ test.describe('Block', () => {
+          await userAModals.unblockWarningModal().unblockButton.click();
+        });
+
+        await test.step('User B receives message sent by User A', async () => {
```
Is it really part of the tests scope to exchange messages between the two users? I'd argue that if A is able to send a message to B we can be sure B isn't blocked. Making sure a blocked user can't send a message is the responsibility of a different test

JacquelineLehner replied:
The test specifications in Testiny require User A and B sending messages to each other

markbrockhoff commented on test/e2e_tests/specs/Block/block.spec.ts:
```diff
@@ -17,83 +17,82 @@
+    let userC: User;
+
+    test.beforeEach(async ({createTeam}) => {
+      const teamA = await createTeam('Team A', {withMembers: 1});
```
Since this runs before each test it would be great to parallelize the team creation:
```suggestion
      const [teamA, teamB] = await Promise.all([createTeam('Team A', {withMembers: 1}), createTeam('Team B')]);
```

markbrockhoff commented on test/e2e_tests/pageManager/webapp/modals/unblockWarning.modal.ts:
```diff
@@ -0,0 +1,48 @@
+
+import {Locator, Page} from '@playwright/test';
+
+export class UnblockWarningModal {
```
There is no component matching this POM, it's just a wrapper around the "modal-template-confirm" modal.
Something like this would allow us to re-use the POM for more than 2 cases ;)

```ts
export class ConfirmModal extends BaseModal {
  constructor(page: Page) {
    super(page, 'modal-template-confirm');
  }
}
```

This already includes all utils to execute the action or click cancel etc, since they are part of the BaseModal

markbrockhoff commented on test/e2e_tests/pageManager/webapp/pages/account.page.ts:
```diff
@@ -43,6 +43,7 @@ export class AccountPage {
   readonly resetPasswordButton: Locator;
   readonly receiveNewsletterCheckbox: Locator;
   readonly typingIndicator: Locator;
+  readonly teamName: Locator;
```
This variable is only used within this file and never used anywhere else -> [YAGNI](https://de.wikipedia.org/wiki/YAGNI)

markbrockhoff commented on test/e2e_tests/pageManager/webapp/pages/conversationList.page.ts:
```diff
@@ -33,6 +35,10 @@ export class ConversationListPage {
   readonly searchConversationsInput: Locator;
   readonly archiveConversationMenuButton: Locator;
   readonly unarchiveConversationMenuButton: Locator;
+  readonly statusLabel: Locator;
```
`statusLabel` and `profilePicture` are initialized but never used -> [YAGNI](https://de.wikipedia.org/wiki/YAGNI)

JacquelineLehner committed:
test(block): refactor block.spec.ts according to Review-Comments

JacquelineLehner committed:
test(block): refactor block.spec.ts according to Review-Comments

e-maad reviewed (dismissed):
LGTM 👍

JacquelineLehner committed:
test(block): fixed unclear comments and variable namings
